packages/jupyter-chat/src/components/input/use-chat-commands.tsx

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) Jupyter Development Team.
+ * Distributed under the terms of the Modified BSD License.
+ */
+
+import React from 'react';
+import { useEffect, useState } from 'react';
+import type {
+  AutocompleteChangeReason,
+  AutocompleteProps
+} from '@mui/material';
+import { Box } from '@mui/material';
+
+import { ChatCommand, IChatCommandRegistry } from '../../chat-commands';
+import { getCurrentWord, replaceCurrentWord } from './utils';
+
+type UseChatCommandsReturn = {
+  autocompleteProps: Omit<AutocompleteProps<any, any, any, any>, 'renderInput'>;
+  menu: {
+    open: boolean;
+    highlighted: boolean;
+  };
+};
+
+/**
+ * A hook which automatically returns the list of command options given the
+ * current input and chat command registry.
+ *
+ * Intended usage: `const chatCommands = useChatCommands(...)`.
+ */
+export function useChatCommands(
+  input: string,
+  setInput: (newInput: string) => void,
+  inputRef: React.MutableRefObject<HTMLInputElement | undefined>,
+  chatCommandRegistry?: IChatCommandRegistry
+): UseChatCommandsReturn {
+  // whether an option is highlighted in the chat commands menu
+  const [highlighted, setHighlighted] = useState(false);
+
+  // whether the chat commands menu is open
+  const [open, setOpen] = useState(true);
+
+  // current list of chat commands matched by the current word.
+  // the current word is the space-separated word at the user's cursor.
+  const [commands, setCommands] = useState<ChatCommand[]>([]);
+
+  useEffect(() => {
+    async function getCommands() {
+      const providers = chatCommandRegistry?.getProviders();
+      const cursorIndex = inputRef.current?.selectionStart;
+      if (!providers || cursorIndex === null || cursorIndex === undefined) {
+        return;
+      }
+
+      const currentWord = getCurrentWord(input, cursorIndex);
+      if (!currentWord?.length) {
+        return;
+      }
+
+      let newCommands: ChatCommand[] = [];
+      for (const provider of providers) {
+        // TODO: optimize performance when this method is truly async
```

Comment:
Maybe this optimization could be perform at the `provider` level.

For example, it could keep in memory the previous word and suggestions. If the new word is only additional letters from the previous one, it only filters the previous suggestions.

Reply:
Agreed! Provider-level optimizations will be necessary for more complex command providers that need to make network requests frequently. However, what I meant by this comment was that we can also do some optimizations in our frontend:

- Use `Promise.all()` to start async `getChatCommands()` tasks in parallel.
- Show chat command suggestions as they resolve, instead of waiting for every provider to finish getting chat commands.

Reply:
Opened a new issue for this: https://github.com/jupyterlab/jupyter-chat/issues/167